src/core/server/server.cc

```diff
@@ -86,6 +88,224 @@
+        });
+  }
+}
+
+void Server::ListenerInterface::LogicalConnection::CancelDrainGraceTimer() {
+  MutexLock lock(&mu_);
+  drain_grace_timer_handle_cancelled_ = true;
+  if (drain_grace_timer_handle_ != EventEngine::TaskHandle::kInvalid) {
+    event_engine_->Cancel(drain_grace_timer_handle_);
+    drain_grace_timer_handle_ = EventEngine::TaskHandle::kInvalid;
+  }
+}
+
+//
+// Server::ListenerInterface
+//
+
+void Server::ListenerInterface::Orphan() {
+  // Cancel the watch before shutting down so as to avoid holding a ref to the
+  // listener in the watcher.
+  if (config_fetcher_watcher_ != nullptr) {
+    CHECK_NE(server_->config_fetcher(), nullptr);
+    server_->config_fetcher()->CancelWatch(config_fetcher_watcher_);
+  }
+  absl::flat_hash_set<OrphanablePtr<LogicalConnection>> connections;
+  {
+    MutexLock lock(&mu_);
+    // Orphan the connections so that they can start cleaning up.
+    connections = std::move(connections_);
+    is_serving_ = false;
+  }
+  // OrphanImpl will unref the listener and hence should be done last.
+  OrphanImpl();
+}
+
+void Server::ListenerInterface::Start() {
+  if (server_->config_fetcher() != nullptr) {
+    auto watcher =
+        std::make_unique<ListenerInterface::ConfigFetcherWatcher>(Ref());
+    config_fetcher_watcher_ = watcher.get();
+    server_->config_fetcher()->StartWatch(
+        grpc_sockaddr_to_string(&resolved_address_, false).value(),
+        std::move(watcher));
+  } else {
+    {
+      MutexLock lock(&mu_);
+      started_ = true;
+      is_serving_ = true;
+    }
+    StartImpl();
+  }
+}
+
+void Server::ListenerInterface::AddLogicalConnection(
+    absl::AnyInvocable<OrphanablePtr<ListenerInterface::LogicalConnection>(
+        const ChannelArgs& args)>
+        connection_creator,
+    const ChannelArgs& args, grpc_endpoint* endpoint) {
+  RefCountedPtr<grpc_server_config_fetcher::ConnectionManager>
+      connection_manager;
+  {
+    MutexLock lock(&mu_);
+    if (!is_serving_) {
+      // Not serving
+      return;
+    }
+    connection_manager = connection_manager_;
+  }
+  ChannelArgs new_args = args;
+  if (server_->config_fetcher() != nullptr) {
+    if (connection_manager == nullptr) {
+      // Connection manager not available
+      return;
+    }
+    absl::StatusOr<ChannelArgs> args_result =
+        connection_manager->UpdateChannelArgsForConnection(new_args, endpoint);
+    if (!args_result.ok()) {
+      return;
+    }
+    auto* server_credentials =
+        (*args_result).GetObject<grpc_server_credentials>();
+    if (server_credentials == nullptr) {
+      // Could not find server credentials
+      return;
+    }
+    auto security_connector =
+        server_credentials->create_security_connector(*args_result);
+    if (security_connector == nullptr) {
+      // Unable to create secure server with credentials
+      return;
+    }
+    new_args = (*args_result).SetObject(security_connector);
+  }
+  auto connection = connection_creator(new_args);
```

Comment:
This is going to call `Start()` on the connection.  In the old code, this happened after adding the connection to `connections_`; in the new code, it's now happening before that.  Is that going to cause any problem?

Reply:
Sorry, you're right, I was misreading this.